compiler/src/service/watch.rs

```diff
@@ -1,47 +1,549 @@
+            // Get the event from the inbox or the watcher.
+            let event = tokio::select! {
+                Some(it) = inbox.recv() => Some(ActorEvent::Message(it)),
+                Some(it) = Self::get_notify_event(&mut self.watcher) => Some(ActorEvent::NotifyEvent(it)),
+                Some(it) = self.undetermined_recv.recv() => Some(ActorEvent::ReCheck(it)),
+            };
+
+            // Failed to get the event.
+            let Some(event) = event else {
+                log::info!("failed to get event, exiting...");
+                return;
+            };
+
+            // Increase the logical tick per event.
+            self.logical_tick += 1;
+
+            // log::info!("vfs-notify event {event:?}");
+            // function entries to handle some event
+            match event {
+                ActorEvent::Message(NotifyMessage::UpstreamUpdate(event)) => {
+                    self.invalidate_upstream(event);
+                }
+                ActorEvent::Message(NotifyMessage::SyncDependency(paths)) => {
+                    if let Some(changeset) = self.update_watches(&paths) {
+                        self.send(FilesystemEvent::Update(changeset));
+                    }
+                }
+                ActorEvent::NotifyEvent(event) => {
+                    // log::info!("notify event {event:?}");
+                    if let Some(event) = log_notify_error(event, "failed to notify") {
+                        self.notify_event(event);
+                    }
+                }
+                ActorEvent::ReCheck(event) => {
+                    self.recheck_notify_event(event).await;
+                }
             }
-            Err(e) => error!("watch error: {:#}", e),
-        },
-        notify::Config::default(),
-    )
-    .map_err(|err| eco_format!("failed to watch directory ({err})"))
-    .unwrap();
-
-    // Add a path to be watched. All files and directories at that path and
-    // below will be monitored for changes.
-    watcher
-        .watch(workspace_dir, RecursiveMode::Recursive)
-        .unwrap();
+        }
+    }
 
-    // Handle events.
-    info!("start watching files...");
-    interrupted_by_events(None);
-    loop {
-        let mut events = vec![];
-        while let Ok(e) =
-            tokio::time::timeout(tokio::time::Duration::from_millis(100), rx.recv()).await
-        {
-            if e.is_none() {
+    /// Update the watches of corresponding invalidation
+    fn invalidate_upstream(&mut self, event: UpstreamUpdateEvent) {
+        // Update watches of invalidated files.
+        let changeset = self.update_watches(&event.invalidates).unwrap_or_default();
+
+        // Send the event to the consumer.
+        self.send(FilesystemEvent::UpstreamUpdate {
+            changeset,
+            upstream_event: Some(event),
+        });
+    }
+
+    /// Update the watches of corresponding files.
+    fn update_watches(&mut self, paths: &[ImmutPath]) -> Option<FileChangeSet> {
```

Comment:
I don't quite understand why `unwatch` + re`watch` works? What problem does it solve?